File: src/algorithms/python/dijkstra.py
import heapq

def dijkstra(grid, start, end):
    """
    Dijkstra's Algorithm for Pathfinding
    Time Complexity: O((V + E) log V) where V is vertices and E is edges
    Space Complexity: O(V)
    
    Dijkstra's algorithm finds the shortest path between nodes in a weighted graph.
    It uses a priority queue to always explore the most promising node first.
    
    Args:
        grid (list): 2D grid representing the map (0 = walkable, 1 = wall)
        start (tuple): Starting position (row, col)
        end (tuple): Target position (row, col)
        
    Returns:
        list: Path from start to end as list of (row, col) tuples, or None if no path exists
    """
    rows, cols = len(grid), len(grid[0])
    
    # Priority queue: stores (distance, row, col)
    pq = [(0, start[0], start[1])]
    
    # Dictionary to store the shortest distance to each cell
    distances = {start: 0}
    
    # Dictionary to reconstruct the path
    came_from = {}
    
    # Directions: up, down, left, right
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while pq:
        current_dist, row, col = heapq.heappop(pq)
        
        # If we reached the end, reconstruct the path
        if (row, col) == end:
            path = []
            current = end
            while current in came_from:
                path.append(current)
                current = came_from[current]
            path.append(start)
            return path[::-1]  # Reverse to get start -> end
        
        # Skip if we've already found a better path to this cell
        if current_dist > distances.get((row, col), float('inf')):
            continue
        
        # Explore all 4 directions
        for dr, dc in directions:
            new_row, new_col = row + dr, col + dc
            
            # Check if the new position is valid
            if (0 <= new_row < rows and 
                0 <= new_col < cols and 
                grid[new_row][new_col] == 0):
                
                # In an unweighted grid, each step costs 1
                new_dist = current_dist + 1
                
                # If we found a shorter path to this cell
                if new_dist < distances.get((new_row, new_col), float('inf')):
                    distances[(new_row, new_col)] = new_dist
                    came_from[(new_row, new_col)] = (row, col)
                    heapq.heappush(pq, (new_dist, new_row, new_col))
    
    # No path found
    return None


# Example usage and test
if __name__ == "__main__":
    # Create a sample grid (0 = walkable, 1 = wall)
    test_grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 0, 0]
    ]
    
    start_pos = (0, 0)
    end_pos = (4, 4)
    
    print(f"Grid size: {len(test_grid)}x{len(test_grid[0])}")
    print(f"Start: {start_pos}, End: {end_pos}")
    
    path = dijkstra(test_grid, start_pos, end_pos)
    
    if path:
        print(f"Path found with {len(path)} steps:")
        print(f"Path: {path}")
        print(f"Total distance: {len(path) - 1}")
    else:
        print("No path found!")
    
    # Test with weighted grid (if you want to extend it)
    print("\nNote: This implementation treats all cells as equal weight (1).")
    print("For weighted grids, modify the grid to store weights instead of 0/1.")

File: src/algorithms/pathfinding/dijkstra.js
/**
 * Copyright (c) 2025 Ayoub Abidi
 * Licensed under Elastic License 2.0 OR Commercial
 * See LICENSE for details.
 */

import { GRID_ELEMENT_STATES } from '../../constants';

/**
 * Dijkstra's Pathfinding Algorithm
 * Time Complexity: O((V + E) log V)
 * Space Complexity: O(V)
 *
 * Finds the shortest path in a weighted graph using a priority queue.
 * For grid pathfinding with uniform weights, behaves similarly to BFS.
 *
 * @param {number[][]} grid - 2D grid with cell weights
 * @param {Object} start - Start position {row, col}
 * @param {Object} end - End position {row, col}
 * @param {number} rows - Number of rows in grid
 * @param {number} cols - Number of columns in grid
 * @returns {Object[]} - Array of animation steps
 */
export function dijkstra(grid, start, end, rows, cols) {
  const steps = [];

  if (!start || !end || !rows || !cols) {
    console.error('Dijkstra: Invalid inputs', { start, end, rows, cols });
    return steps;
  }

  if (
    start.row < 0 ||
    start.row >= rows ||
    start.col < 0 ||
    start.col >= cols
  ) {
    console.error('Dijkstra: Start position out of bounds', {
      start,
      rows,
      cols,
    });
    return steps;
  }

  if (end.row < 0 || end.row >= rows || end.col < 0 || end.col >= cols) {
    console.error('Dijkstra: End position out of bounds', { end, rows, cols });
    return steps;
  }

  const distances = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const visited = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(false));
  const parent = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(null));

  const states = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(GRID_ELEMENT_STATES.DEFAULT));
  states[start.row][start.col] = GRID_ELEMENT_STATES.START;
  states[end.row][end.col] = GRID_ELEMENT_STATES.END;

  steps.push({
    grid: grid.map(row => [...row]),
    states: states.map(row => [...row]),
    description: 'algorithms.descriptions.dijkstra',
  });

  distances[start.row][start.col] = 0;

  const pq = [{ row: start.row, col: start.col, dist: 0 }];

  const directions = [
    { row: -1, col: 0 }, // up
    { row: 1, col: 0 }, // down
    { row: 0, col: -1 }, // left
    { row: 0, col: 1 }, // right
  ];

  let found = false;

  while (pq.length > 0 && !found) {
    pq.sort((a, b) => a.dist - b.dist);
    const current = pq.shift();

    if (visited[current.row][current.col]) {
      continue;
    }

    visited[current.row][current.col] = true;

    if (
      !(current.row === start.row && current.col === start.col) &&
      !(current.row === end.row && current.col === end.col)
    ) {
      states[current.row][current.col] = GRID_ELEMENT_STATES.CLOSED;
    }

    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: `Exploring cell (${current.row}, ${current.col}) with distance ${current.dist.toFixed(1)}`,
    });

    if (current.row === end.row && current.col === end.col) {
      found = true;
      break;
    }

    for (const dir of directions) {
      const newRow = current.row + dir.row;
      const newCol = current.col + dir.col;

      if (newRow < 0 || newRow >= rows || newCol < 0 || newCol >= cols) {
        continue;
      }

      if (visited[newRow][newCol]) {
        continue;
      }

      const newDist = distances[current.row][current.col] + 1;

      if (newDist < distances[newRow][newCol]) {
        distances[newRow][newCol] = newDist;
        parent[newRow][newCol] = current;
        pq.push({ row: newRow, col: newCol, dist: newDist });

        if (!(newRow === end.row && newCol === end.col)) {
          states[newRow][newCol] = GRID_ELEMENT_STATES.OPEN;
        }
      }
    }
  }

  if (found) {
    const path = [];
    let current = end;

    while (
      current &&
      !(current.row === start.row && current.col === start.col)
    ) {
      path.unshift(current);
      current = parent[current.row][current.col];
    }

    for (const cell of path) {
      if (!(cell.row === end.row && cell.col === end.col)) {
        states[cell.row][cell.col] = GRID_ELEMENT_STATES.PATH;
      }
    }

    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: `Path found! Length: ${path.length} cells, Total distance: ${distances[end.row][end.col]}`,
    });
  } else {
    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: 'No path found',
    });
  }

  return steps;
}

/**
 * Pure Dijkstra implementation for testing (returns path coordinates)
 * @param {Object} start - Start position {row, col}
 * @param {Object} end - End position {row, col}
 * @param {number} rows - Number of rows
 * @param {number} cols - Number of columns
 * @returns {Object[]|null} - Array of path coordinates or null if no path
 */
export function dijkstraPure(start, end, rows, cols) {
  const distances = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const visited = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(false));
  const parent = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(null));

  distances[start.row][start.col] = 0;
  const pq = [{ row: start.row, col: start.col, dist: 0 }];

  const directions = [
    { row: -1, col: 0 },
    { row: 1, col: 0 },
    { row: 0, col: -1 },
    { row: 0, col: 1 },
  ];

  while (pq.length > 0) {
    pq.sort((a, b) => a.dist - b.dist);
    const current = pq.shift();

    if (visited[current.row][current.col]) {
      continue;
    }

    visited[current.row][current.col] = true;

    if (current.row === end.row && current.col === end.col) {
      const path = [];
      let curr = end;
      while (curr) {
        path.unshift(curr);
        curr = parent[curr.row][curr.col];
      }
      return path;
    }

    for (const dir of directions) {
      const newRow = current.row + dir.row;
      const newCol = current.col + dir.col;

      if (
        newRow >= 0 &&
        newRow < rows &&
        newCol >= 0 &&
        newCol < cols &&
        !visited[newRow][newCol]
      ) {
        const newDist = distances[current.row][current.col] + 1;
        if (newDist < distances[newRow][newCol]) {
          distances[newRow][newCol] = newDist;
          parent[newRow][newCol] = { row: current.row, col: current.col };
          pq.push({ row: newRow, col: newCol, dist: newDist });
        }
      }
    }
  }

  return null; // No path found
}
